File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        
        if (!root) return nullptr;

        // if both are greater, explore right side
        if (p->val > root->val && q->val > root->val){
            return lowestCommonAncestor(root->right, p, q);
        // if both are less than, explore left side
        } else if (p->val < root->val && q->val < root->val){
            return lowestCommonAncestor(root->left, p, q);
        // if there's a divergence, return the root
        } else {
            return root;
        }

    }
};

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], sub_root: Optional[TreeNode]) -> bool:
        
        if not root: return False
        
        if self.isSameTree(root, sub_root):
            return True
        
        return self.isSubtree(root.left, sub_root) or self.isSubtree(root.right, sub_root)
    
    def isSameTree(self, root: Optional[TreeNode], sub_root: Optional[TreeNode]) -> bool:
        
        # leaf nodes, return True
        if not root and not sub_root: return True   
        
        # structure is different
        if (root and not sub_root) or (not root and sub_root): return False

        # values are different
        if root.val != sub_root.val: return False

        # keep recursing
        return self.isSameTree(root.left, sub_root.left) and self.isSameTree(root.right, sub_root.right)